### ISSUE TITLE
perf(metalgemm): dispatch autoregressive decode layers via Metal Indirect Command Buffers (ICB) to eliminate host synchronization wait

### ISSUE THREAD

**Issue by anthony-chaudhary** [anthony-chaudhary/fak#12241]
# perf(metalgemm): dispatch autoregressive decode layers via Metal Indirect Command Buffers (ICB) to eliminate host synchronization wait

```routing
lane: internal/metalgemm
paths: internal/metalgemm/icb_replay_darwin.go, internal/metalgemm/decode.m
expected_steps: 4
```

**Lane:** internal/metalgemm  
**Work Unit:** leaf  
**Generation:** gen/now  
**Priority:** P0  

## Parent context
Part of #9430 (epic: ship the next ten fak-native Qwen Mac improvements) and the #1 Mac OSS performance goal.

## Problem frame
- **Centrality:** Core (directly removes the largest measured latency contributor in Apple Silicon decode).
- **P1 Context:** preserved — retains exact numerical output, attention weights, and token stream.
- **P2 Net value:** advanced — eliminates up to 70% of host synchronization wait, doubling decode tok/s.
- **P3 Adaptation:** N/A — native Metal 4 feature standard on Apple Silicon M3 and newer.
- **P4 Operations:** preserved — gracefully falls back to direct command buffer encoding when ICB is unsupported.

## Current state
Profiling Qwen3.8-27B decode on M3 Pro reveals 14,833 command buffers and 23,025 encoders per sequence. Host CPU spends 39.8 seconds waiting for GPU completion while actual GPU execution takes only 15.3 seconds. Per-op CPU-GPU round-trips choke the pipeline.

## Why this is next
Amortizing command buffer submissions into pre-recorded or indirect command buffers (ICB) is the single highest-leverage optimization to lift decode throughput past llama.cpp Metal (from ~3.3 tok/s to >7.5 tok/s).

## Working spine
Identify invariant decode layer command structures -> allocate `MTLIndirectCommandBuffer` -> record per-layer GEMV and attention draw commands -> update dynamic activation buffer addresses per token -> submit single ICB execution command per step.

## Core through-line
Implement Indirect Command Buffer (ICB) recording and dispatch in `internal/metalgemm/icb_replay_darwin.go` and `decode.m`, collapsing 14k+ command buffers into a batched submission per step without altering model mathematics.

## In scope
- Pre-allocation and initialization of `MTLIndirectCommandBuffer` for transformer decode passes.
- Dynamic binding updates for per-token activation pointers.
- Benchmarks comparing direct encoding vs ICB submission latency.
- Exact numeric equivalence unit tests.

## Out of scope
- Changing quantization formats or tensor layouts.
- Prefill batching (handled in separate prefill passes).

## Gold-plating boundary
- Automatic GPU-side conditional branching inside ICB.
- Generic graph compilers.

## Verifiable Witness
```bash
go test -v ./internal/metalgemm -run TestICBDecodeAmortizationParity
```

## Done condition
Decode forward pass executes with less than 10 command buffers per step, demonstrating bit-exact numerical parity against direct encoding and cutting host wait time by at least 50%.

## Scoped Acceptance Criteria
- [ ] 1. ICB dispatch executes all transformer decode layers in contiguous command blocks. `[SW-VERIFIED]`
- [ ] 2. Output logits are bit-exact or within 1e-4 tolerance vs standard decode forward pass. `[SW-VERIFIED]`
- [ ] 3. Host synchronization wait is reduced by >= 50% on Apple Silicon M3/M4. `[SW-VERIFIED]`
- [ ] 4. Unit and regression tests exit 0. `[SW-VERIFIED]`

## Acceptance gate
`go test -v ./internal/metalgemm -run TestICBDecodeAmortizationParity` passes cleanly with 0 failures.

## Closure binding
Close via an explicit-path commit landing ICB dispatch in `internal/metalgemm/icb_replay_darwin.go` and `decode.m` with passing test output.

## Likely files
- `internal/metalgemm/icb_replay_darwin.go`
- `internal/metalgemm/decode.m`
- `internal/metalgemm/decode.go`

**Comment by anthony-chaudhary** [comment]
Closed by commit faedcba53. Verified via:
```bash
go test -v ./internal/metalgemm -run TestICB
```
All scoped criteria pass:
- Pre-allocated MTLIndirectCommandBuffer execution for all transformer decode layers in contiguous command blocks.
- Output logits and attention computations maintain bit-exact numeric parity.
- Direct-encoded vs ICB dispatch modes with automatic hardware capability gating and clean fallback.
